segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/Throttler.java

```diff
@@ -171,6 +171,11 @@ boolean isThrottlingRequired() {
             // may find it and act on it.
             val result = new InterruptibleDelay(delayFuture, delay.getDurationMillis(), delay.getThrottlerName());
             this.currentDelay.set(result);
+            //check if we have throtle-exempt operations after calculating a interruptible delay
+            //and just before throttling.
+            if (this.isSuspended.get()) {
+                notifyThrottleSourceChanged();
```

Comment:
Please, let's think how we can create a test to exercise this specific situation.

Reply:
And also, this notifyThrottleSourceChanged is not atomic with the following futures, there is still a small window in between that the new coming throttle-exampt operation form another thread got lost in track and got throttled unexpectedly.

Reply:
> The point here is that does the throttle-exampt operation ever get a chance to be enqueued after the throttler triggered?

When the throttler is triggered, using the throttle method, the first thing that it does is check if there are throttle exempt operations present, if so suspend the throttler, and this happens in an independently running loop in OperationProcessor. But what if a throttle-exempt operation "A" is enqued post this check in the throttle method and the next throttle-exempt operation comes a few seconds later. There is a possibility of  "A" getting throttled for those few seonds. And thats what this PR tries avoiding. Now based on tests done a minimal setup( where throttle-exempt operations were gettting throttled) we would see a very very small percentage of them, like maybe 10-15 out of 30-40k of throttle-exmpt operations, were getting throttled and with this change we dont see even that small number of operations getting throttled.

Reply:
@abhinb  is it possible to synchronize the aforementioned check step and throttle-exempt operation enqueue step? So that we can not miss any throttle-exempt operations.

Reply:
since there is a race between when the delay is cancelled and set, and since by that time,it would mean we already have the thottle-exmpt operation enqueued, won't the suspended check that we have added act as a guard here?

Reply:
@realAaronWu this is in the hot write path. If during the design of it no synchronization has been put in place for this specific point was probably for a reason, and we may need a really strong justification for adding it. Otherwise, we may be inducing an important performance penalty without no gain. In other words, only if there is proof that there is a race condition that could be inducing to some real problem we may consider adding synchronization here (assuming it solve the problem).

Reply:
i suppose these changes can be additive. As in we would not want to introduce any extra synchronization in the path and see how this fares, as this code where the change has been added,  has been without any synch to begin with. In case we see issues, we know from this discussion what areas to look at.

Reply:
@abhinb  thanks, it sounds reasonable, let's keep an eye on the unexpected throttling.

Reply:
@abhinb @RaulGracia Please use the [Local Stress Tester](https://github.com/pravega/pravega/wiki/Local-Stress-Testing) (Self Tester) to validate any changes on the write hot path from a perf POV. You can do it locally on your machine. Set it up to run with in-mem Tier 1 & LTS and run a few benchmarks before and a few after the change (stop any other processes running on the host machine to minimize interference). Test it with small and large events and make sure this throttler is actually exercised. Also, be on the lookout for deadlocks if you want to add additional sync primitives around this path.